### PR TITLE
AmqpTransportInitiator should handle synchronous completion

### PR DESCRIPTION
**Pull request by xinchen10**
... of inner transport connection.

### CODE REVIEW

**Comment by mastermanu** [comment]
LGTM

File: Microsoft.Azure.Amqp/Amqp/Transport/AmqpTransportInitiator.cs [hunk 1]
```diff
@@ -41,8 +41,15 @@ public override bool ConnectAsync(TimeSpan timeout, TransportAsyncCallbackArgs c
...
             this.timeoutHelper = new TimeoutHelper(timeout);
-            innerInitiator.ConnectAsync(timeout, args);
-            return true;
+            if (innerInitiator.ConnectAsync(timeout, args))
+            {
+                return true;
+            }
+            else
+            {
+                this.OnConnectComplete(args);
```
Comment (nayato) [inline Microsoft.Azure.Amqp/Amqp/Transport/AmqpTransportInitiator.cs]: don't know if that's intended here but you're breaking the documented behavior: 
```Returns true if connect is pending, the callback is invoked upon completion.
Returns false if connect is completed synchronously, the callback is not invoked.```
I'd expect no callback to be called here + `ConnectTaskAsync` should deal with false being returned.

File: Microsoft.Azure.Amqp/Amqp/Transport/AmqpTransportInitiator.cs [hunk 2]
```diff
@@ -41,7 +41,12 @@ public override bool ConnectAsync(TimeSpan timeout, TransportAsyncCallbackArgs c
...
             callbackArgs.CompletedSynchronously = false;
             this.timeoutHelper = new TimeoutHelper(timeout);
-            innerInitiator.ConnectAsync(timeout, args);
+            if (!innerInitiator.ConnectAsync(timeout, args))
+            {
+                this.OnConnectComplete(args);
+            }
+
+            // always complete from callback
             return true;
```
Comment (nayato) [inline Microsoft.Azure.Amqp/Amqp/Transport/AmqpTransportInitiator.cs]: why not return `innerInitiator.ConnectAsync(timeout, args)` and handle `false` in `ConnectTaskAsync`?

Comment (xinchen10) [inline Microsoft.Azure.Amqp/Amqp/Transport/AmqpTransportInitiator.cs]: After the transport is connected, we may optionally need to perform transport upgrade (e.g. tcp - ssl) or sasl negotiation, which is also part of the initiator's connect job. We could track the state of each step in those operations but I think that will make the code even more complicated.

Comment (xinchen10) [inline Microsoft.Azure.Amqp/Amqp/Transport/AmqpTransportInitiator.cs]: After the transport is connected, we may optionally need to perform transport upgrade (e.g. tcp - ssl) or sasl negotiation, which is also part of the initiator's connect job. We could track the state of each step in those operations but I think that will make the code even more complicated.

Comment (nayato) [inline Microsoft.Azure.Amqp/Amqp/Transport/AmqpTransportInitiator.cs]: Shouldn't be that different from what socket does, i.e.: if underlying call completes synchronously, perform own "callback work synchronously but do not call callback received from calling sight, just return false.

**Pull request merged**